popdat/src/import_census.rs

```diff
@@ -1,128 +1,113 @@
-        // all are MultiPoly's with just one element, but some truly have multiple polygons.
-        //
-        // When we implement downloading, importer/src/utils.rs has a download() helper that we
-        // could copy here. (And later dedupe, after deciding how this crate will integrate with
-        // the importer)
-        let path = abstutil::path(format!(
-            "system/{}/population_areas.geojson",
-            map.get_name().city
-        ));
-        let bytes = abstutil::slurp_file(&path).map_err(|s| anyhow!(s))?;
-        debug!("parsing geojson at path: {}", &path);
-        timer.start("parsing geojson");
-        let geojson = GeoJson::from_reader(&*bytes)?;
-        timer.stop("parsing geojson");
-        let mut results = vec![];
-        let collection = geojson::FeatureCollection::try_from(geojson)?;
+    pub fn fetch_all_for_map(map: &Map, timer: &mut Timer) -> anyhow::Result<Vec<CensusArea>> {
+        timer.start("processing population areas fgb");
+        let areas = tokio::runtime::Runtime::new()
+            .expect("Failed to create Tokio runtime")
+            .block_on(Self::fetch_all_for_map_async(map, timer))?;
+        timer.stop("processing population areas fgb");
+        Ok(areas)
+    }
+
+    async fn fetch_all_for_map_async(
+        map: &Map,
+        timer: &mut Timer<'_>,
+    ) -> anyhow::Result<Vec<CensusArea>> {
+        use flatgeobuf::HttpFgbReader;
+        use geozero_core::geo_types::Geo;
 
         let map_area = map.get_boundary_polygon();
         let bounds = map.get_gps_bounds();
 
-        use geo::algorithm::map_coords::MapCoordsInplace;
+        use geo::algorithm::{bounding_rect::BoundingRect, map_coords::MapCoordsInplace};
         let mut geo_map_area: geo::Polygon<_> = map_area.clone().into();
         geo_map_area.map_coords_inplace(|c| {
             let projected = geom::Pt2D::new(c.0, c.1).to_gps(bounds);
             (projected.x(), projected.y())
         });
 
-        debug!("collection.features: {}", &collection.features.len());
-        timer.start("converting to `CensusArea`s");
-        for feature in collection.features.into_iter() {
-            let population = match feature
-                .property("population")
-                .ok_or(anyhow!("missing 'population' property"))?
-            {
-                serde_json::Value::Number(n) => n
-                    .as_u64()
-                    .ok_or(anyhow!("unexpected population number: {:?}", n))?
-                    as usize,
-                _ => {
-                    bail!(
-                        "unexpected format for 'population': {:?}",
-                        feature.property("population")
-                    );
+        timer.start("opening FGB reader");
+        // See the import handbook for how to prepare this file.
+        let mut fgb =
+            HttpFgbReader::open("https://s3.amazonaws.com/mjk_asdf/abs/population_areas.fgb")
```

Comment:
it probably makes sense to move this file to the abstreet s3 buckets.

(it's currently 11.5gb FYI)

Reply:
Just finished downloading it, now struggling to re-upload. aws CLI couldn't figure out how to do a direct copy from another public-through-HTTP bucket. :P
It'll eventually be at abstreet.s3-website.us-east-2.amazonaws.com/population_areas.fgb. I don't think sense it makes sense to put it in /dev/data/input or anything like that, because that directory is synced with a local copy. I'm not sure if it makes sense to let people opt into having a local copy of this file through the updater.

Reply:
Yeah I've been thinking about the offline use case and haven't come to any conclusions. Relying on downloading a GBs large file seems fraught. Maybe it'd make sense to instead have the importer generate and cache the final scenario?

Reply:
That might be the easiest route forward. It might be a choppy gameplay experience to have to read from a remote file, do all of the pathfinding to pick modes, etc just to start a scenario. So for most maps, we can precompute this. If we still want to have live tuning of some of the parameters into the scenario generation (which seems likely -- % of students, if census doesn't tell us), then we can figure out some caching for the extracted CensusAreas. Some of it might be a little more interesting on the web.

Reply:
http://abstreet.s3-website.us-east-2.amazonaws.com/population_areas.fgb is now live

Reply:
Thanks - I've updated the link.